spec/lib/crawler/url_validator/url_request_check_spec.rb

```diff
@@ -0,0 +1,199 @@
+    double('UrlCrawlResult', status_code: 200, content_type: 'text/html', duration: 0.5, location: nil, error: nil,
+                             suggestion_message: nil)
+  end
+  let(:url_crawl_result_redirect) do
+    instance_double('Crawler::Data::CrawlResult::Redirect',
+                    redirect?: true,
+                    location: Crawler::Data::URL.parse('http://redirected.com'))
+  end
+  let(:validator) { described_class.new(url: valid_url, crawl_config:) }
+
+  describe '#validate_url_request' do
+    before do
+      allow(validator).to receive(:http_executor).and_return(http_executor)
+      allow(http_executor).to receive(:run).and_return(url_crawl_result)
+    end
+
+    context 'when the status code is 200' do
+      it 'validates successfully' do
+        expect(validator)
+          .to receive(:validation_ok)
+          .with(:url_request,
+                "Successfully fetched #{valid_url}: HTTP 200.",
+                hash_including(:status_code, :content_type, :request_time_msec))
+        validator.validate_url_request
+      end
+    end
+
+    context 'when the status code is 204' do
+      before { allow(url_crawl_result).to receive(:status_code).and_return(204) }
+
+      it 'fails validation with no content message' do
+        expect(validator)
+          .to receive(:validation_fail)
+          .with(:url_request,
+                "The Web server at #{valid_url} returned no content (HTTP 204).",
+                hash_including(:status_code, :content_type, :request_time_msec))
+        validator.validate_url_request
+      end
+    end
+
+    context 'when the status code is a redirect (301, 302, 303, 307, 308)' do
+      before { allow(url_crawl_result).to receive(:status_code).and_return(301) }
+
+      it 'calls redirect_validation_result' do
+        expect(validator).to receive(:redirect_validation_result)
+          .with(hash_including(:status_code, :content_type, :request_time_msec))
+        validator.validate_url_request
+      end
+    end
+
+    context 'when the status code is 401' do
+      before { allow(url_crawl_result).to receive(:status_code).and_return(401) }
+
+      it 'calls unauthorized_validation_result' do
+        expect(validator).to receive(:unauthorized_validation_result)
+          .with(hash_including(:status_code, :content_type, :request_time_msec))
+        validator.validate_url_request
+      end
+    end
+
+    context 'when the status code is 403' do
+      before { allow(url_crawl_result).to receive(:status_code).and_return(403) }
+
+      it 'fails validation with forbidden message' do
+        expect(validator)
+          .to receive(:validation_fail)
+          .with(:url_request,
+                "The web server at #{valid_url} denied us permission to view that page (HTTP 403).\nThis website " \
+                "may require a user name and password.\nRead more at: " \
+                'https://www.elastic.co/guide/en/enterprise-search/current/crawler-managing.html' \
```

Comment:
@navarone-feekery we should probably pull out any references to these docs, and swap them for links to this repo instead. Let's do that in a separate PR.

Reply:
I've made an issue to track this: https://github.com/elastic/crawler/issues/178